File: frontend/src/lib/schemas/integrations.ts
// Mirrors backend/schemas/integrations.py. Keep in sync.
import { z } from "zod";

export const handoffKindSchema = z.enum(["tailor", "skeleton", "save"]);

export const extensionHandoffRequestSchema = z.object({
  kind: handoffKindSchema,
  job_description: z.string().min(1),
  job_url: z.string().optional(),
  company: z.string().optional(),
  role: z.string().optional(),
});

export const extensionHandoffSchema = z.object({
  id: z.string(),
  user_id: z.string(),
  kind: handoffKindSchema,
  job_description: z.string(),
  job_url: z.string().nullable().optional(),
  company: z.string().nullable().optional(),
  role: z.string().nullable().optional(),
  expires_at: z.string().datetime(),
  used_at: z.string().datetime().nullable().optional(),
  created_at: z.string().datetime(),
});

export const extensionHandoffCreatedResponseSchema = z.object({
  handoff_id: z.string(),
  expires_at: z.string().datetime(),
});

export const googleConnectStartSchema = z.object({
  user_id: z.string().min(1),
  return_url: z.string().min(1),
});

export const googleConnectStartResponseSchema = z.object({
  oauth_url: z.string(),
  state: z.string(),
  expires_at: z.string().datetime(),
});

export const googleIntegrationSchema = z.object({
  user_id: z.string(),
  spreadsheet_id: z.string(),
  scopes: z.array(z.string()),
  connected_at: z.string().datetime(),
});

export const googleDisconnectRequestSchema = z.object({
  user_id: z.string().min(1),
});

export type HandoffKind = z.infer<typeof handoffKindSchema>;
export type ExtensionHandoffRequest = z.infer<typeof extensionHandoffRequestSchema>;
export type ExtensionHandoff = z.infer<typeof extensionHandoffSchema>;
export type ExtensionHandoffCreatedResponse = z.infer<
  typeof extensionHandoffCreatedResponseSchema
>;
export type GoogleConnectStart = z.infer<typeof googleConnectStartSchema>;
export type GoogleConnectStartResponse = z.infer<typeof googleConnectStartResponseSchema>;
export type GoogleIntegration = z.infer<typeof googleIntegrationSchema>;
export type GoogleDisconnectRequest = z.infer<typeof googleDisconnectRequestSchema>;

File: backend/schemas/integrations.py
# backend/schemas/integrations.py
"""Extension handoff + Google Sheets integration schemas."""
from datetime import datetime
from typing import List, Literal, Optional
from pydantic import BaseModel, Field


HandoffKind = Literal["tailor", "skeleton", "save"]


class ExtensionHandoffCreate(BaseModel):
    """Sent by the Chrome extension to /api/extension/handoff.

    user_id is derived from the verified HMAC-signed cookie token on the
    server, never trusted from the request body. This schema is what the
    server constructs internally before insert.
    """
    user_id: str = Field(min_length=1)
    kind: HandoffKind
    job_description: str = Field(min_length=1)
    job_url: Optional[str] = None
    company: Optional[str] = None
    role: Optional[str] = None


class ExtensionHandoffRequest(BaseModel):
    """The wire-format request body from the extension. user_id is NOT here;
    it comes from the signed cookie. Validated separately."""
    kind: HandoffKind
    job_description: str = Field(min_length=1)
    job_url: Optional[str] = None
    company: Optional[str] = None
    role: Optional[str] = None


class ExtensionHandoff(BaseModel):
    id: str
    user_id: str
    kind: HandoffKind
    job_description: str
    job_url: Optional[str] = None
    company: Optional[str] = None
    role: Optional[str] = None
    expires_at: datetime
    used_at: Optional[datetime] = None
    created_at: datetime


class ExtensionHandoffCreatedResponse(BaseModel):
    handoff_id: str
    expires_at: datetime


class GoogleConnectStart(BaseModel):
    user_id: str = Field(min_length=1)
    return_url: str = Field(min_length=1)


class GoogleConnectStartResponse(BaseModel):
    oauth_url: str
    state: str
    expires_at: datetime


class GoogleIntegration(BaseModel):
    user_id: str
    spreadsheet_id: str
    scopes: List[str]
    connected_at: datetime


class GoogleDisconnectRequest(BaseModel):
    user_id: str = Field(min_length=1)
